noxfile.py

```diff
@@ -0,0 +1,182 @@
+    """
+
+    to_hash = {}
+    for key, value in metadata.items():
+        if os.path.exists(value):
+            with open(value) as fd:
+                to_hash[key] = fd.read()
+        else:
+            to_hash[key] = value
+
+    to_hash = json.dumps(to_hash).encode("utf8")
+    sha1 = hashlib.sha1(to_hash).hexdigest()
+    sha1_path = os.path.join(session.virtualenv.location, "metadata.sha1")
+
+    if session.virtualenv._reused:
+        if os.path.exists(sha1_path):
+            with open(sha1_path) as fd:
+                should_reinstall = fd.read().strip() != sha1
+        else:
+            should_reinstall = True
+    else:
+        should_reinstall = True
+
+    with open(sha1_path, "w") as fd:
+        fd.write(sha1)
+
+    if should_reinstall:
+        session.debug("updating environment since the dependencies changed")
+
+    return should_reinstall
+
+
+# ==================================================================================== #
+#                              nox sessions definitions                                #
+# ==================================================================================== #
+
+
+for name in EXAMPLES:
+
+    @nox.session(name=name, venv_backend="conda")
+    def example(session, name=name):
+        environment_yml = f"examples/{name}/environment.yml"
+        if should_reinstall_dependencies(session, environment_yml=environment_yml):
+            session.run(
+                "conda",
+                "env",
+                "update",
+                "--prune",
+                f"--file={environment_yml}",
+                f"--prefix={session.virtualenv.location}",
+            )
+
+            # install sphinx-gallery and its dependencies
+            session.install("sphinx-gallery", "sphinx", "pillow", "matplotlib")
+
+        session.run("python", "generate-gallery.py", f"examples/{name}")
+        os.unlink(f"docs/src/examples/{name}/index.rst")
+
+        if "--no-build-docs" not in session.posargs:
+            session.notify("build_docs")
+
+
+@nox.session(venv_backend="none")
+def docs(session):
+    """Run all examples and build the documentation"""
+
+    for example in EXAMPLES:
+        session.run("nox", "-e", example, "--", "--no-build-docs", external=True)
+
+    session.run("nox", "-e", "build_docs", external=True)
+
+
+@nox.session
+def build_docs(session):
+    """Assemble the documentation into a website, assuming pre-generated examples"""
+
+    requirements = "docs/requirements.txt"
+    if should_reinstall_dependencies(session, requirements=requirements):
+        session.install("-r", requirements)
+
+    with open("docs/src/index.rst", "w") as output:
+        with open("docs/src/index.rst.in") as fd:
+            output.write(fd.read())
+
+        output.write(".. toctree::\n")
+        output.write("    :caption: Table of Contents\n")
+        output.write("    :maxdepth: 1\n")
+        output.write("\n")
```

Comment:
This seems to be static. Should we write this directly in `index.rst.in`?

Reply:
👍